pilot/pkg/xds/ads.go

```diff
@@ -340,6 +351,10 @@ func (s *DiscoveryServer) shouldRespond(con *Connection, request *discovery.Disc
 	con.proxy.WatchedResources[request.TypeUrl].LastRequest = request
 	con.proxy.Unlock()
 
+	// Signal waiting semaphore in push operations that a valid
+	// ACK has been processed as determined by matched nonce
+	<-con.semaphore
```

Comment:
How does the first request work? If I understand this, this is the first code called and it is waiting for a message in `con.semaphore`?

Reply:
Also what happens when Istiod sends response and Envoy disconnects before it acks it?

Reply:
For disconnects, it does not matter. When reconnect there will a new connection established

Reply:
@howardjohn I am not all that convinced I understand why this works as it does. You are right that this code is the first code called, and it is waiting for an empty message in `con.semaphore`.

Here is my understanding.

1. The receive thread is started.
1. The receive thread blocks here (or is running to reach this point).
1. A proxy is discovered by pilot, creates a new connection, and starts the XDS push.
1. An event is pushed in one of the two push semaphores in the main thread.
1. The receive thread is waiting for that event - or alternatively, the push (main thread) is waiting for the establishment of the blocking channel in 3. As both channels are blocking, golang guarantees this behavior (afiact:).
1. One of the blocking channels in 3 is signaled, unblocking the receive thread. Blocking on the semaphores in 3 only happens during the startup of the receive thread.
1. My main concern with this code is that if the receive thread were to fail prior to the establishment of the proxy (killed by OOM killer, stray gamma ray, etc) , there may be a permanent failure in the main thread, hence my concerns about edge-triggered, vs level-triggered).

From there things become more straightforward. Each matched nonce blocks, each push unblocks. At this point normal flow control is maintained.

Another concern may be the lack of a matched nonce in the receive thread (Envoy takes a long lunch break). I am not sure if there are healthcheck timers on the envoy connection to prevent this scenario in discovery.go. I should probably take a look.

Even if my understanding is correct (and I'm a pilot noob, so YMMV) - the 3 lines of addition should be guarded with a feature flag, and maybe even test cases could be developed :)

Cheers,
-steve

Reply:
exchange 2 and 3 - wrong ordering.

Reply:
with more understanding of the code, 3 and 6 are not consistent. `processRequest` processes a request from the receive thread via channel from `receive` (duh...). I would have expected this to be named differently (`processReceive`) - but that's fine - now I understand it more but not completely, I can document it 👍 

@howardjohn I did implement @hzxuzhonghu 's approach, and while it works correctly, the chart zigzags after ~230 services as if the semaphore wasn't present. I am not sure why that is the case. Any help here is appreciated, as I'd like to see this merge before the freeze deadline for 1.8.